r2dbc-spec/src/main/asciidoc/connections.adoc

```diff
@@ -0,0 +1,222 @@
+Metadata must not require a connection to a data source.
+
+.`ConnectionFactoryMetadata` Interface
+====
+[source,java]
+----
+public interface ConnectionFactoryMetadata {
+
+    String getName();
+
+}
+----
+====
+
+
+See the R2DBC SPI Specification for more details.
+
+[[connections.factory.discovery]]
+== `ConnectionFactory` Discovery Mechanism
+
+As part of its usage, the `ConnectionFactories` class attempts to load any R2DBC driver classes referenced
+by the `ConnectionFactoryProvider` interface listed in the Java Service Provider manifests available on the CLASSPATH.
+
+Drivers must include the file `META-INF/services/io.r2dbc.spi.ConnectionFactoryProvider`.
+This file contains the name of the R2DBC driver's implementation (or implementations) of `io.r2dbc.spi.ConnectionFactoryProvider`.
+To ensure that drivers can be loaded using this mechanism, `io.r2dbc.spi.ConnectionFactoryProvider` implementations are required to provide a no-argument constructor.
+
+.META-INF/services/io.r2dbc.spi.ConnectionFactoryProvider file contents
+====
+[source]
+----
+com.example.ConnectionFactoryProvider
+----
+====
+
+.`ConnectionFactoryProvider` Interface
+====
+[source,java]
+----
+public interface ConnectionFactoryProvider {
+
+    ConnectionFactory create(ConnectionFactoryOptions connectionFactoryOptions);
+
+    boolean supports(ConnectionFactoryOptions connectionFactoryOptions);
+
+}
+----
+====
+
+`ConnectionFactories` uses a `ConnectionFactoryOptions` object to lookup a matching driver using a two-step model:
+
+1. Lookup of an adequate `ConnectionFactoryProvider`.
+2. Obtain the `ConnectionFactory` from the `ConnectionFactoryProvider`.
+
+`ConnectionFactoryProvider` implementations are required to return a `boolean` indicator whether they support a specific configuration represented by `ConnectionFactoryOptions`.
+Drivers must expect any plurality of ``Option``s to be configured.
+Drivers must report that they support a configuration only if the `ConnectionFactoryProvider` can provide a `ConnectionFactory` based on the given `ConnectionFactoryOptions`.
+Drivers should gracefully fail if a `ConnectionFactory` creation through `ConnectionFactoryProvider.create(…)` is not possible.
+
+See the R2DBC SPI Specification for more details.
+
+[[connections.factory.options]]
+== The `ConnectionFactoryOptions` Class
+
+The `ConnectionFactoryOptions` class represents a configuration to request a `ConnectionFactory` from a `ConnectionFactoryProvider`.
+It represents the <<overview.connection, programmatic connection creation>> approach without using driver-specific classes.
+`ConnectionFactoryOptions` instances are created using the builder pattern and properties are configured through `Option<T>` identifiers.
+A `ConnectionFactoryOptions` is immutable once created.
+`Option` objects are reused as part of the built-in constant pool.
+Options are identified by a literal.
+
+`ConnectionFactoryOptions` defines a set of well-known options:
```

Comment:
Note to myself: Drivers should not rely on the option value type bound by `Option<T>` as generics are not retained during runtime.